Lib/pathlib/types.py

```diff
@@ -72,38 +80,38 @@ class _JoinablePath(ABC):
 
     @property
     @abstractmethod
-    def parser(self):
+    def parser(self) -> _PathParser:
         """Implementation of pathlib._types.Parser used for low-level path
         parsing and manipulation.
         """
         raise NotImplementedError
 
     @abstractmethod
-    def with_segments(self, *pathsegments):
+    def with_segments(self: _JP, *pathsegments: Union[_JP, str]) -> _JP:
```

Comment:
I think pathsegments should be `str` actually. It's not always possible to support joining two objects of the same type, e.g. two `ZipPath` objects with different underlying zip files.

Reply:
This also applies to `joinpath()`, `__truediv__` and `__rtruediv__`

Reply:
changed to `str`